### PR TITLE
IF: Net plugin takes appropriate actions on invalid QC claims and invalid vote messages

### PR DESCRIPTION
**Pull request by linh2931**
When an invalid QC claim or an invalid vote message is received, Net plugin needs to disconnect the peer immediately, keep track of the occurrence, or do nothing, based on the nature of validation failures. This PR classifies validation failures and implements the corresponding actions.

Resolved https://github.com/AntelopeIO/leap/issues/1569.

### CODE REVIEW

File: libraries/chain/include/eosio/chain/hotstuff/hotstuff.hpp [hunk 1]
```diff
@@ -92,6 +92,14 @@ namespace eosio::chain {
       invalid                  // invalid message (other reason)
    };
 
+   enum class vote_status {
+      succeeded,
```
Comment (greg7mdp) [inline libraries/chain/include/eosio/chain/hotstuff/hotstuff.hpp:96]: no big deal, but I think `success` is more typical as a status code than `succeeded`

**Review by greg7mdp — APPROVED** [review]

**Comment by ericpassmore** [comment]
Note:start
group: IF
category: INTERNALS
summary: Take appropriate action when invalid claim or vote is received. 
Note: end

File: plugins/net_plugin/net_plugin.cpp [hunk 2]
```diff
@@ -261,13 +261,17 @@ namespace eosio {
       bool verify_catchup( const connection_ptr& c, uint32_t num, const block_id_type& id ); // locks mutex
 
    public:
+      enum class closing_mode {
+         immediately,
+         back_off
```
Comment (heifner) [inline plugins/net_plugin/net_plugin.cpp:266]: This `closing_mode` is confusing to me. It is used to determine if you close or handshake. 
Not sure what `backoff` implies. It currently means: don't close but rather send a handshake.

File: plugins/net_plugin/net_plugin.cpp [hunk 3]
```diff
@@ -910,6 +914,7 @@ namespace eosio {
       std::atomic<bool>               block_sync_throttling{false};
       std::atomic<std::chrono::nanoseconds>   last_bytes_sent{0ns};
       std::atomic<boost::asio::ip::port_type> remote_endpoint_port{0};
+      std::atomic<uint64_t>           invalid_finality_msg_total{0};
```
Comment (heifner) [inline plugins/net_plugin/net_plugin.cpp:917]: This appears to not be used. It is only incremented and set to 0, never checked. Seems like the `block_status_monitor` is doing everything you need.
We may decide we want to increase `max_consecutive_rejected_windows_` if we find we see a lot of `vote_status::unknown_block`. Or we may wish to track them separately.

File: plugins/net_plugin/net_plugin.cpp [hunk 4]
```diff
@@ -3691,8 +3701,22 @@ namespace eosio {
...
+      switch( cc.process_vote_message(msg) ) {
+         case vote_status::success:
+            my_impl->bcast_vote_message(connection_id, msg);
+            break;
+         case vote_status::unknown_public_key:
+         case vote_status::invalid_signature: // close peer immediately
+            close( false ); // do not reconnect after closing
+            break;
+         case vote_status::unknown_block: // keep tracking
+            ++invalid_finality_msg_total; // atomic
```
Comment (heifner) [inline plugins/net_plugin/net_plugin.cpp:3714]: Should this call `c->block_status_monitor_.rejected();` ?

**Review by heifner — CHANGES_REQUESTED** [review]

File: plugins/net_plugin/net_plugin.cpp (hunk 3 above)
Comment (linh2931) [reply plugins/net_plugin/net_plugin.cpp:917]: https://github.com/AntelopeIO/leap/issues/1569 specifies to track the failures and indicates eventually we may decide to disconnect. The counter is implemented for that purpose. Will ask @arhag in today's meeting to clarify.

Comment (linh2931) [reply plugins/net_plugin/net_plugin.cpp]: Use block_status_monitor per discussion. Thanks.

**Review by heifner — APPROVED** [review]

**Pull request merged**